File: src/main/scala/com/leetcode/TwoSum.scala
package com.leetcode

object TwoSum {

  def twoSum(nums: Array[Int], target: Int): Array[Int] = {
    def iter (index: Int, map: Map[Int, Int]): Array[Int] = {
      if (index == nums.length) {
        throw new IllegalArgumentException
      } else {
        val elem = nums(index)
        val secTerm = target - elem
        if (map.contains(secTerm)) {
          Array(map(secTerm), index)
        } else {
          iter(index + 1, map + (elem -> index))
        }
      }
    }

    iter(0, Map())
  }
}

File: src/main/java/com/leetcode/ds/array/TwoSum.java
package com.leetcode.ds.array;

import java.util.HashMap;
import java.util.Map;

public class TwoSum {
    private TwoSum() {
    }

    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>(nums.length);
        for (int i = 0; i < nums.length; i++) {
            map.put(nums[i], i);
        }
        for (int i = 0; i < nums.length; i++) {
            int diff = target - nums[i];
            Integer diffIndex = map.get(diff);
            if (diffIndex != null && diffIndex != i) {
                return new int[]{i, diffIndex};
            }
        }
        return new int[]{};
    }
}
